docs(docs-infra): Throw build error on invalid guide links

JeanMeche opened the pull request:
The list of valid links is generated from the guides configuration in the ADEV app.
Redirections are knowingly exclude so we stop referencing them.

This commit also fixes all necessary links.


This is to detect issues like #66246 at build time.

josephperrott commented on adev/shared-docs/pipeline/shared/marked/transformations/heading.mts:
```diff
@@ -9,29 +9,30 @@
   // extract the extended markdown heading id
   // ex:  ## MyHeading {# myId}
+  // This is recommended in case we end up having duplicate Ids but we still want the same heading text.
+  // We don't want to make Id generation statful/too complex to handle duplicates automatically.
```
```suggestion
  // We don't want to make Id generation stateful/too complex to handle duplicates automatically.
```

josephperrott commented on adev/shared-docs/pipeline/shared/marked/renderer.mts:
```diff
@@ -17,7 +17,18 @@ import {codespanRender} from './transformations/code.mjs';
+   */
   markdownFilePath?: string;
+
+  /**
```
```suggestion
   */
markdownFilePath?: string;
  /**
```

josephperrott commented on adev/shared-docs/pipeline/guides/index.mts:
```diff
@@ -21,33 +21,45 @@ interface ApiManifestPackage {
+  const apiManifestStr = await readFile(apiManifestPath, {encoding: 'utf8'});
+  apiManifest = JSON.parse(apiManifestStr);
+
+  let parsedRoutes: string[] = [];
```
We should just use the same name here

```suggestion
  let definedRoutes: string[] = [];
```

josephperrott commented on adev/shared-docs/pipeline/shared/marked/renderer.mts:
```diff
@@ -46,41 +57,36 @@ export class AdevDocsRenderer extends Renderer {
+
+    if (
+      !this.context.definedRoutes ||
+      this.context.definedRoutes.length === 0 ||
```
This check should be at the top to shortcut all of the other logic that might need to be considered.

```
    if (!(this.context.definedRoutes?.length)) {
      return true;
    }
```

josephperrott commented on adev/shared-docs/pipeline/_guides.bzl:
```diff
@@ -85,6 +92,10 @@ generate_guides = rule(
             doc = """A file containing API entries to be used in the markdown.""",
             allow_single_file = True,
         ),
+        "definedRoutes": attr.label(
```
Through convention, the names for bazel attributes on rules shoud be snake cased


```suggestion
        "defined_routes": attr.label(
```

JeanMeche commented on adev/src/app/routing/navigation-entries/index.ts:
```diff
@@ -0,0 +1,1670 @@
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import type {NavigationItem} from '@angular/docs';
```
It's more like, this is now we file we need to update when we add routes. `adev/src/app/routing/sub-navigation-data.ts` is almost empty now

JeanMeche committed:
docs(docs-infra): Throw build error on invalid guide links

The list of valid links is generated from navigation data configuration in the ADEV app.
Redirections are knowingly exclude so we stop referencing them.

This commit also fixes all necessary links.

josephperrott commented on adev/shared-docs/pipeline/guides/index.mts:
```diff
@@ -21,33 +21,45 @@ interface ApiManifestPackage {
         markdownFilePath: filePath,
         apiEntries: mapManifestToEntries(apiManifest),
         highlighter,
+        definedRoutes: definedRoutes,
```
```suggestion
        definedRoutes,
```

josephperrott reviewed (approved):
LGTM

JeanMeche committed:
fixup! docs(docs-infra): Throw build error on invalid guide links

thePunderWoman commented:
This PR was merged into the repository. The changes were merged into the following branches:

- main: c09f5352f670ffc37929c1b1095ca01236621f92
- 21.1.x: 79f979a673aed56664f951fac1d5659b91e1d016
